File: plugins/dynamics/transient-shaper/transient-shaper.js
// Transient Shaper — independent control of attack and sustain. Two
// parallel envelope followers chase the input: one fast, one slow. The
// difference between them = the transient component (mostly nonzero
// right after a hit and decaying quickly). The slow envelope = the
// sustain component. Multiply each by user-controllable gain and sum.
// SPL Transient Designer reborn — fatten snares, soften plosive Ps,
// add bite to a vocal, take edges off a guitar.

function TransientShaperProcessor() {
  this.fastEnv = 0;
  this.slowEnv = 0;
  this.fastGain = 1;
  this.slowGain = 1;
}

TransientShaperProcessor.parameterDescriptors = [
  { name: 'attack',  label: 'Attack',  defaultValue: 0,   minValue: -1, maxValue: 1 },
  { name: 'sustain', label: 'Sustain', defaultValue: 0,   minValue: -1, maxValue: 1 },
  { name: 'mix',     label: 'Mix',     defaultValue: 1,   minValue: 0,  maxValue: 1 }
];

TransientShaperProcessor.prototype.process = function (inputs, outputs, parameters) {
  var input = inputs[0][0], output = outputs[0][0];
  if (!input || !output) return true;
  var attack = parameters.attack[0];
  var sustain = parameters.sustain[0];
  var mix = parameters.mix[0];

  // Two envelope followers — same attack speed (fast), different release.
  // The fast envelope falls quickly, so it tracks the transient peak;
  // the slow one lingers, so it represents the body of the note.
  var attCoef = 1 - Math.exp(-1 / Math.max(1, sampleRate * 0.002));
  var fastRelCoef = 1 - Math.exp(-1 / Math.max(1, sampleRate * 0.030));
  var slowRelCoef = 1 - Math.exp(-1 / Math.max(1, sampleRate * 0.200));

  // Map -1..+1 → gain multipliers (0.0 .. 4.0). -1 cuts the component
  // completely, 0 leaves it as-is, +1 quadruples it.
  var attGain = attack >= 0 ? 1 + attack * 3 : 1 + attack * 0.95;
  var susGain = sustain >= 0 ? 1 + sustain * 3 : 1 + sustain * 0.95;

  for (var i = 0; i < input.length; i++) {
    var x = input[i];
    var rect = x < 0 ? -x : x;

    var fc = rect > this.fastEnv ? attCoef : fastRelCoef;
    this.fastEnv = this.fastEnv + fc * (rect - this.fastEnv);
    var sc = rect > this.slowEnv ? attCoef : slowRelCoef;
    this.slowEnv = this.slowEnv + sc * (rect - this.slowEnv);

    // Transient signal = how much the fast envelope is currently above
    // the slow one (a positive bump right after each onset).
    var trans = this.fastEnv - this.slowEnv;
    if (trans < 0) trans = 0;
    // Convert to a multiplier riding over 1.0. With unit gains, output
    // equals input; with attack>0, transients are boosted, etc.
    var origAmp = this.slowEnv > 1e-6 ? this.slowEnv : 1e-6;
    var transMul = 1 + (attGain - 1) * (trans / (origAmp + trans));
    var susMul = susGain;
    // Apply transient multiplier proportional to where in the envelope
    // we are. Simplest blend: gain = susMul during sustain region,
    // gain = susMul * transMul at the transient peak. Sample-by-sample
    // weight = trans / (trans + slow).
    var w = trans / (trans + origAmp);
    var gain = susMul * (1 - w) + susMul * transMul * w;
    var wet = x * gain;
    output[i] = x * (1 - mix) + wet * mix;
  }
  return true;
};

registerProcessor('transient-shaper', TransientShaperProcessor);

File: plugins/dynamics/transient-shaper/TransientShaper.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Transient Shaper (native port) — independent attack/sustain via the
// difference of fast and slow envelopes. SPL Transient Designer in
// spirit — fatten snares, soften plosives, add bite to vocals.
public final class TransientShaper implements VocalMonitorNativePlugin {
    private float fastEnv = 0f, slowEnv = 0f;
    private int sampleRate = 44100;
    private float attack = 0f, sustain = 0f, mix = 1f;

    @Override
    public void init(int sr) { this.sampleRate = sr; fastEnv = 0f; slowEnv = 0f; }

    @Override public String[] parameterNames() { return new String[] { "attack", "sustain", "mix" }; }
    @Override public float parameterMin(String n) { return "mix".equals(n) ? 0f : -1f; }
    @Override public float parameterMax(String n) { return 1f; }
    @Override public float parameterDefault(String n) {
        switch (n) {
            case "attack":  return 0f;
            case "sustain": return 0f;
            case "mix":     return 1f;
            default:        return 0f;
        }
    }
    @Override public String parameterLabel(String n) {
        switch (n) {
            case "attack":  return "Attack";
            case "sustain": return "Sustain";
            case "mix":     return "Mix";
            default:        return n;
        }
    }
    @Override public void setParameter(String n, float v) {
        switch (n) {
            case "attack":  attack = v; break;
            case "sustain": sustain = v; break;
            case "mix":     mix = v; break;
        }
    }

    @Override
    public void process(float[] input, float[] output) {
        final float attCoef = 1f - (float) Math.exp(-1.0 / Math.max(1, sampleRate * 0.002));
        final float fastRelCoef = 1f - (float) Math.exp(-1.0 / Math.max(1, sampleRate * 0.030));
        final float slowRelCoef = 1f - (float) Math.exp(-1.0 / Math.max(1, sampleRate * 0.200));
        final float attGain = attack >= 0 ? 1f + attack * 3f : 1f + attack * 0.95f;
        final float susGain = sustain >= 0 ? 1f + sustain * 3f : 1f + sustain * 0.95f;
        final float mixLocal = mix;
        final float dry = 1f - mixLocal;
        float fE = fastEnv, sE = slowEnv;
        final int n = input.length;
        for (int i = 0; i < n; i++) {
            float x = input[i];
            float rect = x < 0 ? -x : x;
            float fc = rect > fE ? attCoef : fastRelCoef;
            fE = fE + fc * (rect - fE);
            float sc = rect > sE ? attCoef : slowRelCoef;
            sE = sE + sc * (rect - sE);
            float trans = fE - sE;
            if (trans < 0f) trans = 0f;
            float origAmp = sE > 1e-6f ? sE : 1e-6f;
            float transMul = 1f + (attGain - 1f) * (trans / (origAmp + trans));
            float w = trans / (trans + origAmp);
            float g = susGain * (1f - w) + susGain * transMul * w;
            float wet = x * g;
            output[i] = x * dry + wet * mixLocal;
        }
        fastEnv = fE; slowEnv = sE;
    }
}
